src/mlpack/methods/neighbor_search/spill_search_rules_impl.hpp

```diff
@@ -0,0 +1,440 @@
+template<typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType,
+         typename SortPolicy,
+         typename MetricType>
+inline double NeighborSearchRules<SortPolicy, MetricType, tree::SpillTree<
+    MetricType, StatisticType, MatType, HyperplaneType, SplitType>>::
+    CalculateBound(TreeType& queryNode) const
+{
+  // This is an adapted form of the B(N_q) function in the paper
+  // ``Tree-Independent Dual-Tree Algorithms'' by Curtin et. al.; the goal is to
+  // place a bound on the worst possible distance a point combination could have
+  // to improve any of the current neighbor estimates.  If the best possible
+  // distance between two nodes is greater than this bound, then the node
+  // combination can be pruned (see Score()).
+
+  // There are a couple ways we can assemble a bound.  For simplicity, this is
+  // described for nearest neighbor search (SortPolicy = NearestNeighborSort),
+  // but the code that is written is adapted for whichever SortPolicy.
+
+  // First, we can consider the current worst neighbor candidate distance of any
+  // descendant point.  This is assembled with 'worstDistance' by looping
+  // through the points held by the query node, and then by taking the cached
+  // worst distance from any child nodes (Stat().FirstBound()).  This
+  // corresponds roughly to B_1(N_q) in the paper.
+
+  double worstDistance = SortPolicy::BestDistance();
+
+  // Loop over points held in the node.
+  for (size_t i = 0; i < queryNode.NumPoints(); ++i)
+  {
+    const double distance = candidates[queryNode.Point(i)].top().first;
+    if (SortPolicy::IsBetter(worstDistance, distance))
+      worstDistance = distance;
+  }
+
+  // Loop over children of the node, and use their cached information to
+  // assemble bounds.
+  for (size_t i = 0; i < queryNode.NumChildren(); ++i)
+  {
+    const double firstBound = queryNode.Child(i).Stat().FirstBound();
+
+    if (SortPolicy::IsBetter(worstDistance, firstBound))
+      worstDistance = firstBound;
+  }
+
+  // At this point, worstDistance holds the value of B_1(N_q).
+
+  // Now consider the parent bounds.
+  if (queryNode.Parent() != NULL)
+  {
+    // The parent's worst distance bound implies that the bound for this node
+    // must be at least as good.  Thus, if the parent worst distance bound is
+    // better, then take it.
+    if (SortPolicy::IsBetter(queryNode.Parent()->Stat().FirstBound(),
+        worstDistance))
+      worstDistance = queryNode.Parent()->Stat().FirstBound();
+  }
+
+  // Could the existing bounds be better?
+  if (SortPolicy::IsBetter(queryNode.Stat().FirstBound(), worstDistance))
+    worstDistance = queryNode.Stat().FirstBound();
+
+  // Cache bounds for later.
+  queryNode.Stat().FirstBound() = worstDistance;
+
+  worstDistance = SortPolicy::Relax(worstDistance, epsilon);
+
+  return worstDistance;
```

Comment:
I believe that `CalculateBound()` is the only function that needs to be changed in the rules class; is that right?  Maybe it is better to use some standalone static method to calculate the bound for spill trees, and then something like this in `NeighborSearchRules::CalculateBound()`:

```
if (IsSpillTree<TreeType>::value)
  return CalculateSpillBound(queryNode);
```

If there are other things that differ in this class, too, it might be necessary to use this same trick there.  I see this kind of as a variant of the `#ifdef` style of making things compatible, so it's kind of hackish, but I think maybe it is better than replicating the entire class in order to change only a little functionality.  What do you think?